### ISSUE TITLE
[expr.pre] p4 Does indirection a pointer value of a base class type that points to a derived class object result in UB

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#261]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[expr.pre] p4 says
> If during the evaluation of an expression, the result is not mathematically defined or **not in the range of representable values for its type, the behavior is undefined**.

Assume the width of `int` is `32` and the width of `char` is 8. Consider such two contract examples:
````cpp
int i = 1;
char r = *reinterpret_cast<char*>(&i); // well-defined
````
Since we strictly defined the range of representable values of `int` and `char`, the value `1` of type `int` can be represented in an object of type `char`, hence it is well-formed. 
````cpp
int i = 999999;
char r = *reinterpret_cast<char*>(&i); // undefined behavior
````
The value `999999` cannot be represented in an object of type `char`, hence the behavior is undefined. 

However, we do not define what the range of representable values of a user-defined type is. So, how to determine whether [expr.pre] p4 applies or not? 

Consider these examples:
````cpp
struct B{};
struct D:B{};
D d;
auto r = *reinterpret_cast<B*>(&d); // points to `d` with type `pointer to B`, UB or not? 
````

````cpp
struct B{
   char b;
};
struct D:B{
   char arr[10];
};
D d;
auto r = *reinterpret_cast<B*>(&d); // points to `d` with type `pointer to B`, UB or not? 
````

**Comment by jensmaurer** [comment]
The statement in [expr.pre] p4 is mostly intended for arithmetic operations such as multiplication or division.  Once pointers get involved, see [basic.lval] p11 for the aliasing rules.
```
int i = 1;
char r = *reinterpret_cast<char*>(&i); // well-defined
```
This is not well-defined; it's well-formed, but reads an unspecified value from the object representation of i, and only because `char` gets a special exception (for example, the value you read will be different on big-endian and little-endian machines).

Your class cases are violating [basic.lval] p11, because the type of the lvalue is "B", but the dynamic type of the object is "D".

**Comment by xmh0511** [comment]
> This is not well-defined; it's well-formed, but reads an unspecified value from the object representation of i, and only because char gets a special exception (for example, the value you read will be different on big-endian and little-endian machines).

The result pointer value does not point to the object of type `char`. It still points to the original object, as per [expr.static.cast] p14
> Otherwise, if the original pointer value points to an object a, and there is an object b of type similar to T that is pointer-interconvertible with a, the result is a pointer to b. **Otherwise, the pointer value is unchanged by the conversion**.  

the object of `char` and the object of `int` are not pointer-convertible. Hence, the pointer value still points to the object of type `int`, and the we get the result `*p` as per [expr.unary.op] p1
> The unary * operator performs indirection. Its operand shall be a prvalue of type “pointer to T”, where T is an object or function type. **The operator yields an lvalue of type T denoting the object or function to which the operand points**.  

So, the result is an lvalue that denotes the object of type `int` and whose type is `char`.

**Comment by jensmaurer** [comment]
For accessing the object representation in particular, there are known bugs that are addressed by in-flight papers; see
https://stackoverflow.com/questions/63724182/can-you-access-the-object-representation-of-any-object-through-a-char
for details.

**Comment by xmh0511** [comment]
> Your class cases are violating [basic.lval] p11, because the type of the lvalue is "B", but the dynamic type of the object is "D".

Find a bug between the definition of dynamic type and [basic.lval] p11, see https://github.com/cplusplus/CWG/issues/262

**Comment by xmh0511** [comment]
So, the remaining issues are: 
> 1. The statement in [expr.pre] p4 is mostly intended for arithmetic operations such as multiplication or division. We should embody the intent in the wording.
> 2. Your class cases are violating [basic.lval] p11, because the type of the lvalue is "B", but the dynamic type of the object is "D". As discussed in https://github.com/cplusplus/CWG/issues/262#issuecomment-1469166994, [basic.lval] p11 only applies to the objects of scalar types due to "access".

**Comment by jensmaurer** [comment]
> [basic.lval] p11 only applies to the objects of scalar types due to "access", so dereferencing the pointer value pointing to the class type does not violate [basic.lval] p11.

Yes, but the subsequent invocation of the copy constructor (to initialize "r") does the bad access on scalars.

**Comment by xmh0511** [comment]
> > [basic.lval] p11 only applies to the objects of scalar types due to "access", so dereferencing the pointer value pointing to the class type does not violate [basic.lval] p11.
> 
> Yes, but the subsequent invocation of the copy constructor (to initialize "r") does the bad access on scalars.

Consider this example
````cpp
struct Base{
   char a;
   Base(Base const& other){  // #0
        a = other.a; // access the scalar object
   }
};
struct D:Base{
   char d;
   D(D const& other){
     // no access to the scalar object
  }
};
int main(){
   D d;
   auto r = *reinterpret_cast<B*>(&d);  / #1
}
````
The copy constructor of `Base` is called on the object `D` at `#1`. However, there is no wording that says which object `other` at `#0` refers to, if that `other` refers to `D`, then `D::d` and `Base::a` have the same type, if that other refers to `Base`, there is no problem at all.
